src/java.base/share/classes/jdk/internal/misc/ThreadFlock.java

```diff
@@ -0,0 +1,594 @@
+        if (permit != newValue) {
+            return (boolean) PERMIT.getAndSet(this, newValue);
+        } else {
+            return newValue;
+        }
+    }
+
+    /**
+     * Throws WrongThreadException if the current thread is not the owner.
+     */
+    private void ensureOwner() {
+        if (Thread.currentThread() != owner())
+            throw new WrongThreadException("Current thread not owner");
+    }
+
+    /**
+     * Throws WrongThreadException if the current thread is not the owner
+     * or a thread contained in the flock.
+     */
+    private void ensureOwnerOrContainsThread() {
+        Thread currentThread = Thread.currentThread();
+        if (currentThread != owner() && !containsThread(currentThread))
+            throw new WrongThreadException("Current thread not owner or thread in flock");
+    }
+
+    /**
+     * Opens a new thread flock. The flock is owned by the current thread. It can be
+     * named to aid debugging.
+     *
+     * <p> This method captures the current thread's {@linkplain ScopeLocal scope-local}
+     * bindings for inheritance by threads created in the flock.
+     *
+     * <p> For the purposes of containment, monitoring, and debugging, the parent
+     * of the new flock is determined as follows:
+     * <ul>
+     * <li> If the current thread is the owner of open flocks then the most recently
+     * created, and open, flock is the parent of the new flock. In other words, the
+     * <em>enclosing flock</em> is the parent.
+     * <li> If the current thread is not the owner of any open flocks then the
+     * parent of the new flock is the current thread's flock. If the current thread
+     * was not started in a flock then the new flock does not have a parent.
+     * </ul>
+     *
+     * @param name the name of the flock, can be null
+     * @return a new thread flock
+     */
+    public static ThreadFlock open(String name) {
+        var flock = new ThreadFlock(name);
+        flock.container.push();
+        return flock;
+    }
+
+    /**
+     * {@return the name of this flock or {@code null} if unnamed}
+     */
+    public String name() {
+        return name;
+    }
+
+    /**
+     * {@return the owner of this flock}
+     */
+    public Thread owner() {
+        return container.owner();
+    }
+
+    /**
+     * Starts the given unstarted thread in this flock.
+     *
+     * <p> The thread is started with the scope-local bindings that were captured
+     * when opening the flock. The bindings must match the current thread's bindings.
+     *
+     * <p> This method may only be invoked by the flock owner or threads {@linkplain
+     * #containsThread(Thread) contained} in the flock.
+     *
+     * @param thread the unstarted thread
+     * @return the thread, started
+     * @throws IllegalStateException if this flock is shutdown or closed
+     * @throws jdk.incubator.concurrent.StructureViolationException if the current
```

Comment:
```suggestion
     * @throws WrongThreadException if the current thread is not the owner or a thread
     * contained in the flock
     * @throws jdk.incubator.concurrent.StructureViolationException if the current
```

Reply:
Thanks, I think this slipped in when moving the SL API to the incubator module. It should be IllegalThreadStateException if the thread has already started, and WrongThreadException if not the owner (or flock).